art/defences/trainer/ibp_certified_trainer_pytorch.py

```diff
@@ -0,0 +1,462 @@
+    | Paper link: https://arxiv.org/pdf/1810.12715.pdf
+    """
+
+    def __init__(
+        self,
+        classifier: "IBP_CERTIFIER_TYPE",
+        nb_epochs: Optional[int] = 20,
+        bound: float = 0.1,
+        loss_weighting: float = 0.1,
+        batch_size: int = 32,
+        use_certification_schedule: bool = True,
+        certification_schedule: Optional[Any] = None,
+        use_loss_weighting_schedule: bool = True,
+        loss_weighting_schedule: Optional[Any] = None,
+        augment_with_pgd: bool = False,
+        pgd_params: Optional["PGDParamDict"] = None,
+    ) -> None:
+        """
+        Create an :class:`.AdversarialTrainerCertified` instance.
+
+        Default values are for MNIST in pixel range 0-1.
+
+        :param classifier: Classifier to train adversarially.
+        :param pgd_params: A dictionary containing the specific parameters relating to regular PGD training.
+                           If not provided, we will default to typical MNIST values.
+                           Otherwise must contain the following keys:
+
+                           * *eps*: Maximum perturbation that the attacker can introduce.
+                           * *eps_step*: Attack step size (input variation) at each iteration.
+                           * *max_iter*: The maximum number of iterations.
+                           * *batch_size*: Size of the batch on which adversarial samples are generated.
+                           * *num_random_init*: Number of random initialisations within the epsilon ball.
+        :param bound: The perturbation range for the interval. If the default certification schedule is used
+                      will be the upper limit.
+        :param loss_weighting: Weighting factor for the certified loss.
+        :param nb_epochs: Number of training epochs.
+        :param use_certification_schedule: If to use a training schedule for the certification radius.
+        :param certification_schedule: Schedule for gradually increasing the certification radius. Empirical studies
+                                       have shown that this is often required to achieve best performance.
+                                       Either True to use the default linear scheduler,
+                                       or a class with a .step() method that returns the updated bound every epoch.
+        :param batch_size: Size of batches to use for certified training.
+        """
+        from art.estimators.certification.interval.pytorch import PyTorchIBPClassifier
+
+        if not isinstance(classifier, PyTorchIBPClassifier):
+            raise ValueError(
+                "The classifier to pass in should be of type PyTorchIBPClassifier which can be found in "
+                "art.estimators.certification.interval.pytorch.PyTorchIBPClassifier"
+            )
+
+        super().__init__(classifier=classifier)
+        self._classifier: "IBP_CERTIFIER_TYPE"
```

Comment:
Is the `_` required here?

Reply:
Changed with additional method in the Trainer class.